### PR TITLE
challenge contract

### PR DESCRIPTION
**Pull request by shelmesky**

### CODE REVIEW

File: contracts/sources/chanllenge.move [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+        r = r + ((*Vector::borrow(v, 5) << 16) as u64);
+        r = r + ((*Vector::borrow(v, 4) << 24) as u64);
+        r = r + ((*Vector::borrow(v, 3) << 32) as u64);
+        r = r + ((*Vector::borrow(v, 2) << 40) as u64);
+        r = r + ((*Vector::borrow(v, 1) << 48) as u64);
+        r = r + ((*Vector::borrow(v, 0) << 56) as u64);
+        r
+    }
+
+    fun empty_address(): address {
```
Comment (lerencao) [inline contracts/sources/chanllenge.move:66]: you can use @0x0 directly

File: contracts/deps/contract_account/sources/contract_account.move [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+            move_to(signer, GenesisSignerCapability{cap});
+        }
+
+        public fun get_contract_signer(): signer acquires GenesisSignerCapability {
+            let cap = borrow_global<GenesisSignerCapability>(CONTRACT_ACCOUNT);
+            Account::create_signer_with_cap(&cap.cap)
+        }
+
+        public fun init_contract_account() {
+            let genesis_account = Account::create_genesis_account(CONTRACT_ACCOUNT);
```
Comment (lerencao) [inline contracts/deps/contract_account/sources/contract_account.move:21]: 这个方法不是这样用的。genesis_account 只能是 0x1

Comment (shelmesky) [reply contracts/deps/contract_account/sources/contract_account.move:21]: thanks，又重新分析了一下代码，好像没发现检查地址必须是 0x1 的地方？maybe i was wrong?

Comment (lerencao) [reply contracts/deps/contract_account/sources/contract_account.move:21]: Timestamp::assert_genesis();

Comment (shelmesky) [reply contracts/deps/contract_account/sources/contract_account.move:21]: sorry，我看到了

File: contracts/sources/chanllenge.move [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+        let step_number = get_step_number(challenge_id);
+
+        let contract_signer = ContractAccount::get_contract_signer();
+        let challenges = borrow_global_mut<Challenges>(Signer::address_of(&contract_signer));
+        let c = Vector::borrow_mut(&mut challenges.value, challenge_id);
+
+        assert!(c.challenger != @0x0, ERR_INVALID_CHALLENGE);
+        assert!(c.challenger == Signer::address_of(signer), ERR_MUST_BE_CHALLENGER);
+
+        let state = Table::borrow(&c.asserted_state, step_number);
```
Comment (lerencao) [inline contracts/sources/chanllenge.move:224]: Table::borrow will raise err if the key is not found

Comment (shelmesky) [reply contracts/sources/chanllenge.move]: problem has fixed: use a function to check key exists in table, if not then throw an error.

```move
fun borrow_table_value(table: &Table::Table<u64, HashValue>, key: u64): &HashValue {
    assert!(Table::contains(table, key), ERR_TABLE_KEY_NOT_EXISTS);
    Table::borrow(table, key)
}
```

**Comment by lerencao** [comment]
I'm gonna close this as a similar pr was merged already

**Pull request closed without merging**